File: dotnet/DesignPatterns/BehavioralDesignPatterns/IteratorDesignPattern/IteratorDesignPattern.Example2/OrderIterator.cs
namespace IteratorDesignPattern.Example2;

public class OrderIterator : IIterator<Order>
{
    private readonly OrderCollection _collection;
    private int _currentIndex = 0;

    public OrderIterator(OrderCollection collection)
    {
        _collection = collection;
    }

    public bool HasNext()
    {
        return _currentIndex < _collection.Count;
    }

    public Order Next()
    {
        if (!HasNext())
        {
            throw new InvalidOperationException("No more orders in collection.");
        }

        return _collection[_currentIndex++];
    }

    public void Reset()
    {
        _currentIndex = 0;
    }
}

File: java/DesignPatterns/BehavioralDesignPatterns/IteratorDesignPattern/IteratorDesignPattern.Example2/src/main/java/behavioral/iterator/example2/OrderIterator.java
package behavioral.iterator.example2;

public class OrderIterator implements IIterator<Order> {
    private final OrderCollection collection;
    private int currentIndex = 0;

    public OrderIterator(OrderCollection collection) { this.collection = collection; }

    @Override
    public boolean hasNext() { return currentIndex < collection.getCount(); }

    @Override
    public Order next() { return collection.get(currentIndex++); }

    @Override
    public void reset() { currentIndex = 0; }
}
